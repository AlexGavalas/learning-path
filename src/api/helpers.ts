import axios from 'axios';

export const fetchFileFromStorage = async (
    filename: string,
): Promise<string> => {
    const fileServerUrl = String(process.env.PUBLIC_FILE_SERVER_URL);

    const response = await axios.get(`${fileServerUrl}/${filename}`, {
        responseType: 'text',
    });

    return response.data as string;
};
