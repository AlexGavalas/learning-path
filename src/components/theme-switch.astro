---
import Button from '~components/button.astro';
import {
    DEFAULT_THEME,
    THEME,
    DARK_IMG_PROPS,
    LIGHT_IMG_PROPS,
} from '~constants';

const themeCookie = Astro.cookies.get('theme');
const theme = themeCookie?.value ?? DEFAULT_THEME;

const attributes = theme === THEME.LIGHT ? LIGHT_IMG_PROPS : DARK_IMG_PROPS;
---

<theme-switch>
    <form method="get">
        <input type="hidden" name="theme" value="toggle" />
        <Button variant="wrapper">
            <img {...attributes} width="20" height="20" />
        </Button>
    </form>
</theme-switch>

<script>
    import { getInitialTheme, setThemeInPage } from '~lib/theme';
    import { THEME, DARK_IMG_PROPS, LIGHT_IMG_PROPS } from '~constants';
    import { updateAttributes } from '~lib/element';

    class ThemeSwitch extends HTMLElement {
        buttonElement: HTMLButtonElement | null;
        imageElement: HTMLImageElement | null;
        theme: string;

        constructor() {
            super();

            this.buttonElement = this.querySelector('button');
            this.imageElement = this.querySelector('img');
            this.theme = getInitialTheme();
            this.setTheme(this.theme);
        }

        setTheme = (newTheme: string) => {
            setThemeInPage(newTheme);

            const attributes =
                newTheme === THEME.LIGHT ? LIGHT_IMG_PROPS : DARK_IMG_PROPS;

            if (this.imageElement) {
                updateAttributes({ element: this.imageElement, attributes });
            }

            this.theme = newTheme;
        };

        toggleTheme = (e: MouseEvent) => {
            this.setTheme(
                this.theme === THEME.LIGHT ? THEME.DARK : THEME.LIGHT,
            );

            e.preventDefault();
        };

        connectedCallback() {
            this.buttonElement?.addEventListener('click', this.toggleTheme);
        }

        disconnectedCallback() {
            this.buttonElement?.removeEventListener('click', this.toggleTheme);
        }
    }

    customElements.define('theme-switch', ThemeSwitch);
</script>
