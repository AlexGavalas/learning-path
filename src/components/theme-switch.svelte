<script>
    import Button from '~components/button.svelte';
    import { DEFAULT_THEME, STORAGE_KEY, THEME } from '~constants';

    const LIGHT_IMG_PROPS = {
        src: '/moon.svg',
        alt: 'Moon',
    };

    const DARK_IMG_PROPS = {
        src: '/sun.svg',
        alt: 'Sun',
    };

    let theme =
        typeof localStorage !== 'undefined'
            ? localStorage.getItem(STORAGE_KEY.THEME) ?? DEFAULT_THEME
            : DEFAULT_THEME;

    const toggleTheme = () => {
        const newTheme = theme === THEME.LIGHT ? THEME.DARK : THEME.LIGHT;

        document.documentElement.classList.toggle(
            'dark',
            newTheme === THEME.DARK,
        );

        document.documentElement.style.setProperty('color-scheme', newTheme);

        localStorage.setItem(STORAGE_KEY.THEME, newTheme);

        theme = newTheme;
    };

    $: image = theme === THEME.LIGHT ? LIGHT_IMG_PROPS : DARK_IMG_PROPS;
</script>

<Button on:click={toggleTheme} variant="wrapper">
    <img src={image.src} alt={image.alt} width={20} height={20} />
</Button>
