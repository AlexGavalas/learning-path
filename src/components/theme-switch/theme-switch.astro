---
import Button from '~components/button/button.astro';
import {
    DEFAULT_THEME,
    THEME,
    ONE_YEAR,
    DARK_IMG_PROPS,
    LIGHT_IMG_PROPS,
} from '~constants';

const themeCookie = Astro.cookies.get('theme');
const theme = themeCookie?.value ?? DEFAULT_THEME;

const attributes = theme === THEME.LIGHT ? LIGHT_IMG_PROPS : DARK_IMG_PROPS;
---

<theme-switch>
    <form method="post" action="/api/theme-toggle">
        <input type="hidden" name="theme" value="toggle" />
        <Button variant="wrapper">
            <img {...attributes} width="20" height="20" />
        </Button>
    </form>
</theme-switch>

<script
    is:inline
    define:vars={{
        DARK_IMG_PROPS,
        LIGHT_IMG_PROPS,
        ONE_YEAR,
        THEME,
    }}
>
    const getThemeFromCookie = () => {
        const themeCookie = document.cookie
            .split('; ')
            .find((row) => row.startsWith('theme='));

        return themeCookie?.split('=')[1];
    };

    const getThemeFromOsPreference = () => {
        const themes = [
            { colorScheme: 'light', theme: THEME.LIGHT },
            { colorScheme: 'dark', theme: THEME.DARK },
        ];

        for (const { colorScheme, theme } of themes) {
            const hasPreference = window.matchMedia(
                `(prefers-color-scheme: ${colorScheme})`,
            ).matches;

            if (hasPreference) {
                return theme;
            }
        }
    };

    const setThemeInPage = (theme) => {
        document.documentElement.classList.toggle(
            'light',
            theme === THEME.LIGHT,
        );

        document.documentElement.classList.toggle('dark', theme === THEME.DARK);

        document.documentElement.style.setProperty('color-scheme', theme);
    };

    const getInitialTheme = () =>
        getThemeFromCookie() ?? getThemeFromOsPreference() ?? DEFAULT_THEME;

    // Set the theme inline to avoid a flash of the default theme
    setThemeInPage(getInitialTheme());

    class ThemeSwitch extends HTMLElement {
        themeChannel =
            typeof BroadcastChannel === 'function'
                ? new BroadcastChannel('learning-path-theme')
                : null;

        buttonElement;
        imageElement;
        theme;

        setCookie = (theme) => {
            document.cookie = `theme=${theme}; max-age=${ONE_YEAR}; path=/;`;
        };

        setTheme = (theme) => {
            setThemeInPage(theme);

            if (this.imageElement) {
                const attributes =
                    theme === THEME.LIGHT ? LIGHT_IMG_PROPS : DARK_IMG_PROPS;

                Object.entries(attributes).forEach(([key, value]) => {
                    this.imageElement.setAttribute(key, value);
                });
            }

            this.setCookie(theme);
            this.theme = theme;
        };

        toggleTheme = (e) => {
            const theme = this.theme === THEME.LIGHT ? THEME.DARK : THEME.LIGHT;

            if (document.startViewTransition) {
                document.startViewTransition(() => {
                    this.setTheme(theme);
                });
            } else {
                this.setTheme(theme);
            }

            this.themeChannel?.postMessage({ theme });

            e.preventDefault();
        };

        handleOsPreferenceChange = (e) => {
            const theme = e.matches ? THEME.DARK : THEME.LIGHT;

            this.setTheme(theme);
        };

        handleBroadcastMessage = (e) => {
            const { theme } = e.data;

            this.setTheme(theme);
        };

        connectedCallback() {
            this.buttonElement = this.querySelector('button');
            this.imageElement = this.querySelector('img');

            this.setTheme(getInitialTheme());

            this.buttonElement?.addEventListener('click', this.toggleTheme);

            // Listem for OS preference changes
            window
                .matchMedia('(prefers-color-scheme: dark)')
                .addEventListener('change', this.handleOsPreferenceChange);

            this.themeChannel?.addEventListener(
                'message',
                this.handleBroadcastMessage,
            );
        }

        disconnectedCallback() {
            this.buttonElement?.removeEventListener('click', this.toggleTheme);

            window
                .matchMedia('(prefers-color-scheme: dark)')
                .removeEventListener('change', this.handleOsPreferenceChange);

            this.themeChannel?.removeEventListener(
                'message',
                this.handleBroadcastMessage,
            );
        }
    }

    customElements.define('theme-switch', ThemeSwitch);
</script>
