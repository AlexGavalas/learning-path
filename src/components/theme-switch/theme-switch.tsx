import { useEffect, useState } from 'react';

import { Button } from '~components/button';
import { DEFAULT_THEME, STORAGE_KEY, THEME } from '~constants';

const LIGHT_IMG_PROPS = {
    src: '/moon.svg',
    alt: 'Moon',
};

const DARK_IMG_PROPS = {
    src: '/sun.svg',
    alt: 'Sun',
};

export const ThemeSwitch = (): JSX.Element | null => {
    const [isMounted, setIsMounted] = useState(false);

    const [theme, setTheme] = useState(
        typeof localStorage !== 'undefined'
            ? localStorage.getItem(STORAGE_KEY.THEME) ?? DEFAULT_THEME
            : DEFAULT_THEME,
    );

    useEffect(() => {
        setIsMounted(true);
    }, []);

    const toggleTheme = (): void => {
        const newTheme = theme === THEME.LIGHT ? THEME.DARK : THEME.LIGHT;

        setTheme(newTheme);

        document.documentElement.classList.toggle(
            'dark',
            newTheme === THEME.DARK,
        );

        document.documentElement.style.setProperty('color-scheme', newTheme);

        localStorage.setItem(STORAGE_KEY.THEME, newTheme);
    };

    const image = theme === THEME.LIGHT ? LIGHT_IMG_PROPS : DARK_IMG_PROPS;

    if (!isMounted) {
        return null;
    }

    return (
        <Button onClick={toggleTheme} variant="wrapper">
            <img {...image} width={20} height={20} />
        </Button>
    );
};
