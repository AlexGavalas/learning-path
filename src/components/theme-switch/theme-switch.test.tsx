import { axe } from 'jest-axe';

import { renderWithUser, screen } from '~test/helpers';

import { ThemeSwitch } from './theme-switch';

describe('<ThemeSwitch />', () => {
    it('renders', () => {
        const { container } = renderWithUser(<ThemeSwitch />);

        expect(container).toMatchSnapshot();
    });

    it('is accessible', async () => {
        const { container } = renderWithUser(<ThemeSwitch />);

        const a11yResults = await axe(container);

        expect(a11yResults).toHaveNoViolations();
    });

    describe('when user toggles theme', () => {
        it('changes theme', async () => {
            const { user } = renderWithUser(<ThemeSwitch />);

            const button = screen.getByRole('button', { name: 'Moon' });

            await user.click(button);

            const buttonImage = screen.getByRole('img');

            expect(buttonImage).toHaveAttribute('alt', 'Sun');

            await user.click(button);

            expect(buttonImage).toHaveAttribute('alt', 'Moon');
        });
    });
});
