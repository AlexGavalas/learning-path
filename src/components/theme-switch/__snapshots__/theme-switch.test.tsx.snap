// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ThemeSwitch /> renders 1`] = `
<div>
  <button
    class="flex items-center rounded border-0 p-2 bg-transparent hover:bg-slate-300 dark:hover:bg-slate-600"
  >
    <img
      alt="Moon"
      height="20"
      src="/moon.svg"
      width="20"
    />
  </button>
</div>
`;
